### PR TITLE
[flang][OpenMP]Add support for OpenMP ERROR directive

### PR DESCRIPTION
**Pull request by Leporacanthicus**
Lowering leads to a TODO, with a test to confirm.

Also testing unparse.

### CODE REVIEW

File: flang/include/flang/Parser/parse-tree.h [hunk 1]
```diff
@@ -4008,6 +4015,13 @@ struct OmpMapClause {
   std::tuple<MODIFIERS(), OmpObjectList, bool> t;
 };
 
+// Ref: [5.2:217-218]
+// message-clause ->
+//    MESSAGE("message-text")
+struct OmpMessageClause {
+  WRAPPER_CLASS_BOILERPLATE(OmpMessageClause, std::string);
```
Comment (kparzysz) [inline flang/include/flang/Parser/parse-tree.h:4022]: The message doesn't have to be a string literal.  It needs to be known at compile time only if there is AT clause with COMPILATION as the argument.

Also, we should be able to accept this:
```
character(*), parameter :: s = "haha"
!$omp error at(compilation) message(s)
```

Comment (Leporacanthicus) [reply flang/include/flang/Parser/parse-tree.h]: Ok, so I have updated to allow pretty much anything there. I think we need some semantics checks to make sure it's not `cos(exp(pi * cmplx(3.0, -2.0))` or something, but for the purposes of compiler doesn't say "I have no idea what this means", it should work.

**Review by kparzysz — APPROVED** [review]
Looks good!  Could you add checks for `-fdebug-dump-parse-tree-no-sema` as well to your testcase?

**Comment by Leporacanthicus** [comment]
> Looks good! Could you add checks for `-fdebug-dump-parse-tree-no-sema` as well to your testcase?

Done. I had to make a couple of minor changes to print the actual value of the AT and SEVERITY.

**Review by kparzysz — APPROVED** [review]
LGTM, thanks!

**Comment by Leporacanthicus** [comment]
I'm not entirely sure if I'm missing something here, or this is just my first "real construct" that I've added, so I had to add it to the flangOmpReport tool.

**Comment by kparzysz** [comment]
There are 3 targets that cause such trouble: `flangFeatureList` `flangOmpReport` `flangPrintFunctionNames`.  You'd wait hours for the linux build to start, then it fails in 3 seconds because one of those failed..  grr..

**Pull request merged**